### ISSUE TITLE
solrfalr public_url not working

### ISSUE THREAD

**Issue by farconada** [beechit/fal_securedownload#40]
Hello

I've configure tx_solr with 
`
plugin.tx_solr.index.queue._FILES.default.filePublicUrl = public_url
plugin.tx_solr.index.queue._FILES.default.url = public_url
`
but url are generated wrong
`
/typo3/index.php?ajaxID=FalSecuredownload%3A%3ApublicUrl&eID=dumpFile&t=f&f=6622&token=61756eccb7
`
I'm using commit 435efaae881dfc901f9a598c87d19f31cc0fd210 (latest) of fal_securedownload

![captura de pantalla de 2016-06-28 13 51 55](https://cloud.githubusercontent.com/assets/469221/16414381/9d1d6dee-3d37-11e6-83e2-03acaf5d8cb3.png)
![captura de pantalla de 2016-06-28 13 41 22](https://cloud.githubusercontent.com/assets/469221/16414382/9d1f5172-3d37-11e6-8ee0-5efc8d0cead1.png)
![captura de pantalla de 2016-06-28 13 29 45](https://cloud.githubusercontent.com/assets/469221/16414383/9d234976-3d37-11e6-80f4-3c9fff159811.png)

Is it a bug or am I doing something wrong?

thanks

**Comment by fsaris** [comment]
Hi,

Config looks okay. Could you debug the `fileMetaDataRetrieved` method to see if it is called? That method is responsible for adding the correct publicUrl.

Kind regards,
Frans

**Comment by tottt** [comment]
Same problem here.
Testet if `SolrFalAspect::fileMetaDataRetrieved` is called. It is indeed. When i change some metadata there it takes effect. But the public_url still isn't correct.

**Comment by fsaris** [comment]
You added the typoscript as described in the documentation?

**Comment by tottt** [comment]
After reinstalling all extensions and reverting back to the default config it all works now.
Must have been the config I guess. Thank You.

**Comment by frans-beech-it** [comment]
You welcome :)

**Comment by LeoniePhiline** [comment]
Just to add a note to people who find this bug searching for similar search terms:

On multi sites, default solr creates a fake frontend environment via `\ApacheSolrForTypo3\Solr\Domain\Index\IndexService::initializeHttpServerEnvironment()`. It sets `$_SERVER['HTTP_HOST']` to the domain record of the item's root page.

solrfal (at least solrfal 4.2 :) ) does **not** do that!

Therefore, `\BeechIt\FalSecuredownload\Aspects\SolrFalAspect::fileMetaDataRetrieved()` is called correctly, but with the wrong `HTTP_HOST` if the file target is not collected from the first of multiple sites in the TYPO3 page tree.

E.g. when executing the solrfal indexer task from the backend, which had been requested as `https://hostname1/typo3/`, then the `dumpfile` links will use `hostname1` even if the files were collected in `hostname2` site context.

Workaround:

In `\ApacheSolrForTypo3\Solrfal\Indexing\DocumentFactory::createDocumentForQueueItem()` add before any other instruction:

`$this->initializeHttpServerEnvironment($item);`

Further below, add that method:

```
    /**
     * @param Item $item
     */
    protected function initializeHttpServerEnvironment(Item $item)
    {
        static $hosts = [];
        $rootpageId = $item->getContext()->getSite()->getRootPageId();
        $hostFound = !empty($hosts[$rootpageId]);

        if (!$hostFound) {
            $rootline = BackendUtility::BEgetRootLine($rootpageId);
            $host = BackendUtility::firstDomainRecord($rootline);
            $hosts[$rootpageId] = $host;
        }

        $_SERVER['HTTP_HOST'] = $hosts[$rootpageId];

        // needed since TYPO3 7.5
        GeneralUtility::flushInternalRuntimeCaches();
    }
```

Now, all public URLs will be created with the correct `HTTP_HOST` taken from the file index queue item's `context -> site -> rootPageId`

**This is a solrfal bug, not a fal_securedownload bug!**

**Comment by frans-beech-it** [comment]
@LeoniePhiline tnx for this extensive investigation. Will forward this to the devs of EXT:solrfal  (cc: @timohund)

**Comment by LeoniePhiline** [comment]
PS: DKD might argue that they do replace the HTTP host in ` \ApacheSolrForTypo3\Solrfal\Service\FieldProcessingService::addTypoScriptFieldsToDocument()` - but that does not help because `\BeechIt\FalSecuredownload\Aspects\SolrFalAspect::fileMetaDataRetrieved()` is called long before that point.

`fal_securedwownload` could create its own workaround by masking HTTP_HOST inside `\BeechIt\FalSecuredownload\Aspects\SolrFalAspect::fileMetaDataRetrieved()`, where it also has access to `\ApacheSolrForTypo3\Solrfal\Queue\Item`.

So - you two can fight about whose bug it is! :)
There's two possible places to fix it.

**Comment by frans-beech-it** [comment]
Lets first see if this can be fixed in ext:solffal else we will add a "fix" here

**Comment by LeoniePhiline** [comment]
Here's how you'd fix it in fal_securedownload:

```diff
diff --git a/Classes/Aspects/SolrFalAspect.php b/Classes/Aspects/SolrFalAspect.php
index 7823d4c..ab7fa55 100644
--- a/Classes/Aspects/SolrFalAspect.php
+++ b/Classes/Aspects/SolrFalAspect.php
@@ -10,6 +10,7 @@ namespace BeechIt\FalSecuredownload\Aspects;
 
 use ApacheSolrForTypo3\Solrfal\Queue\Item;
 use BeechIt\FalSecuredownload\Security\CheckPermissions;
+use TYPO3\CMS\Backend\Utility\BackendUtility;
 use TYPO3\CMS\Core\Resource\File;
 use TYPO3\CMS\Core\SingletonInterface;
 use TYPO3\CMS\Core\Utility\ArrayUtility;
@@ -64,7 +65,44 @@ class SolrFalAspect implements SingletonInterface
 
         // Re-generate public url
         $this->publicUrlAspect->setEnabled(false);
+
+        // Remember original http host value
+        $originalHttpHost = isset($_SERVER['HTTP_HOST']) ? $_SERVER['HTTP_HOST'] : null;
+        $this->initializeHttpServerEnvironment($item);
+
+        // Are there cases where we would also need to create (and later restore) a fake TSFE?
+
         $metadata['public_url'] = $item->getFile()->getPublicUrl();
+
+        // restore http host
+        if (!is_null($originalHttpHost)) {
+            $_SERVER['HTTP_HOST'] = $originalHttpHost;
+        } else {
+            unset($_SERVER['HTTP_HOST']);
+        }
+
         $this->publicUrlAspect->setEnabled(true);
     }
+
+
+    /**
+     * @param Item $item
+     */
+    protected function initializeHttpServerEnvironment(Item $item)
+    {
+        static $hosts = [];
+        $rootpageId = $item->getContext()->getSite()->getRootPageId();
+        $hostFound = !empty($hosts[$rootpageId]);
+
+        if (!$hostFound) {
+            $rootline = BackendUtility::BEgetRootLine($rootpageId);
+            $host = BackendUtility::firstDomainRecord($rootline);
+            $hosts[$rootpageId] = $host;
+        }
+
+        $_SERVER['HTTP_HOST'] = $hosts[$rootpageId];
+
+        // needed since TYPO3 7.5
+        GeneralUtility::flushInternalRuntimeCaches();
+    }
 }
```

**Comment by timohund** [comment]
@fsaris thanks for the mention, from a short look on the code in solrfal something like this is allready done:

The method FieldProcessingService::initializeHttpHost, is setting the host from the first domain record. But by now only when realurl is loaded. Maybe this is the problem. I think we need to investigate that.

@LeoniePhiline  are you using realurl? 

Maybe it is just an option to do this also when realurl is not installed. I am asking that since i want to avoid to do the same checks and calls twice or more.

**Comment by LeoniePhiline** [comment]
@timohund There's several issues with that

1) Indeed `realurl`. It is absolutely unnecessary to depend on realurl just to get a page rootline's first domain record. Just check my code at https://github.com/beechit/fal_securedownload/issues/40#issuecomment-400700335 or the `tx_solr` source code.
That's a design flaw in `tx_solrfal` that you can easily correct.

2) I think you haven't read https://github.com/beechit/fal_securedownload/issues/40#issuecomment-400696062 : 
Yes, `solrfal` _does_ indeed use `\ApacheSolrForTypo3\Solrfal\Service\FieldProcessingService::initializeHttpHost()` inside `\ApacheSolrForTypo3\Solrfal\Service\FieldProcessingService::addTypoScriptFieldsToDocument()` - but that's far too late.

Just look at how `tx_solr` does it: 

When it ***starts*** indexing an item - namely in `\ApacheSolrForTypo3\Solr\Domain\Index\IndexService::indexItem()`, the proper environment is set up for the indexing. 
It is a very well known TYPO3-problem that you cannot reliably build frontend URIs without creating a fake environment for `typolink` and in many cases set up an entire fake `TypoScriptFrontendController`.

So the _latest possible_ place where solrfal should do this is in `\ApacheSolrForTypo3\Solrfal\Indexing\DocumentFactory::createDocumentForQueueItem()`! Because already in that method, `\ApacheSolrForTypo3\Solrfal\Indexing\DocumentFactory::addFileInformation()` is calling `getPublicUrl()` in many cases, and thus creating frontend URIs.

Creating the proper environment (`HTTP_HOST` et. al.) in  `\ApacheSolrForTypo3\Solrfal\Indexing\DocumentFactory::addFieldsFromTypoScriptConfiguration()` is **far** too late, since most information to be indexed has already been gathered at this point.

Therefore, what you should do, @timohund , is to wrap the instructions of `\ApacheSolrForTypo3\Solrfal\Indexing\DocumentFactory::createDocumentForQueueItem()` in a construct like https://github.com/beechit/fal_securedownload/issues/40#issuecomment-400700335 does with its `initializeHttpServerEnvironment(Item $item)`.

And `\ApacheSolrForTypo3\Solrfal\Service\FieldProcessingService::initializeHttpHost()` of course must be scrapped. Neither is it defined or called in the right place, nor is it using the right logic - being dependent on realurl.

**Comment by timohund** [comment]
After resolving some side effects of the provided patch, this issue is fixed in solrfal 5.1.0 . Since no feedback on my test request was given i see the issue as resolved from my perspective.

**Comment by lorenzulrich** [comment]
> After resolving some side effects of the provided patch, this issue is fixed in solrfal 5.1.0 . Since no feedback on my test request was given i see the issue as resolved from my perspective.

Is the 4.x branch out of maintenance? I have the same problem with Solr 7/Solrfal 4.2.

Thanks @LeoniePhiline for investigating and providing a patch solution.
